merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/generator/MissionModelGenerator.java

```diff
@@ -646,6 +658,100 @@ public JavaFile generateActivityTypes(final MissionModelRecord missionModel) {
+                    .addMember("value", "$S", MissionModelProcessor.class.getCanonicalName())
+                    .build())
+            .addModifiers(Modifier.PUBLIC, Modifier.FINAL);
+
+    record ComponentMapperNamePair(String componentName, String mapperName) {}
+    for (final var record : recordTypes) {
+      if (record.getKind() != ElementKind.RECORD) {
+        messager().printError("@%s is only allowed on records".formatted(AutoValueMapper.class.getSimpleName()), record);
+      }
+
+      final var methodName = record.toString().replace(".", "_");
+      final var componentToMapperName = new ArrayList<ComponentMapperNamePair>();
+      final var necessaryMappers = new HashMap<TypeMirror, String>();
+      for (final var element : record.getEnclosedElements()) {
+        if (!(element instanceof RecordComponentElement el)) continue;
+        final var typeMirror = el.getAccessor().getReturnType();
+        final var elementName = element.toString();
+        final var valueMapperIdentifier = typeMirror.toString().replace(".", "_") + "ValueMapper";
+        componentToMapperName.add(new ComponentMapperNamePair(elementName, valueMapperIdentifier));
+        necessaryMappers.put(typeMirror, valueMapperIdentifier);
+      }
+
+      final var typeRule = new TypeRule(
+          new TypePattern.ClassPattern(ClassName.get(ValueMapper.class), List.of(TypePattern.from(record.asType()))),
+          Set.of(),
+          necessaryMappers
+              .keySet()
+              .stream()
+              .map(component -> (TypePattern) new TypePattern.ClassPattern(
+                  ClassName.get(ValueMapper.class),
+                  List.of(new TypePattern.ClassPattern((ClassName) ClassName.get(component).box(), List.of()))))
+              .toList(),
+          typeName,
+          methodName);
+      typeRules.add(typeRule);
+
+      final var methodBuilder = MethodSpec
+          .methodBuilder(methodName)
+          .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+          .returns(ParameterizedTypeName.get(ClassName.get(ValueMapper.class), TypeName.get(record.asType())))
+          .addParameters(
+              necessaryMappers
+                  .entrySet()
+                  .stream()
+                  .map(mapperRequest -> ParameterSpec
+                      .builder(
+                          ParameterizedTypeName.get(
+                              ClassName.get(ValueMapper.class),
+                              ClassName.get(mapperRequest.getKey()).box()),
+                          mapperRequest.getValue(),
+                          Modifier.FINAL)
+                      .build())
+                  .toList());
+
+      final var iter = componentToMapperName.iterator();
+      final var codeBlockBuilder = CodeBlock.builder();
+      codeBlockBuilder.add("return new $T<>(\n", RecordValueMapper.class);
+      codeBlockBuilder.add("  $L.class,\n", record.toString());
```

Comment:
Instead of using `toString` (which is not always guaranteed to produce a meaningful name on `TypeElement`s), we should use JavaPoet's static `ClassName.get(TypeElement)` method. We could also use `TypeElement`'s own `getQualifiedName()`, which returns a `Name` (a kind of `CharSequence`, yes Java is weird), but `ClassName` works better with codegen -- which brings me to the next point:

As a rule of thumb, we should use `$T` whenever the literal refers to a type. For instance, JavaPoet can use the fully-qualified class name if the short name is already used by another class in scope. It just gives JavaPoet more to work with.

```suggestion
      codeBlockBuilder.add("  $T.class,\n", ClassName.get(record));
```

Reply:
📏 👍